### ISSUE TITLE
Initialize ggml_cgraph on heap

### ISSUE THREAD

**Issue by LoganDark** [ggerganov/ggml#251]
Since ggml doesn't support iteration or changing shapes in graphs, I find myself having to encode [ridiculous things](https://github.com/saharNooby/rwkv.cpp/blob/c41ed9896b6c1309fcd3f6b8307f34a184bb188a/rwkv.cpp#L1026-L1034) into the computation graph.

Because of this, I hit GGML_MAX_NODES very easily and frequently, to the point where I've had to raise it to 5000000 (5 million) just to get the thing to run properly without hitting an assert.

The problem, of course, is that GGML's public API encourages passing the `ggml_cgraph` on the stack:

https://github.com/ggerganov/ggml/blob/f52d2a05cf8327baf6c0d49e7b231953179e03d3/include/ggml/ggml.h#L1106

I get around this by [initializing it manually on the heap, then using only `ggml_build_forward_expand`](https://github.com/saharNooby/rwkv.cpp/blob/c41ed9896b6c1309fcd3f6b8307f34a184bb188a/rwkv.cpp#L1347-L1350), which allows me to avoid ever having it touch the stack, and therefore handle millions of nodes with ease.

But this is not an officially supported use case, is it? ggml could change at any time and break this, and I'd have to replicate even more implementation details if cgraph initialization ever gets more complex.

So I'm opening this issue to track the addition (or support) of an official solution to this problem.

As far as I can tell, there's no way to prevent me needing such a large computation graph, because running a single graph multiple times is many times slower than simply encoding all the steps into a single graph. (Maybe because ggml messes with threads even when n_threads == 1)

So the next step is to make it easier to manage very large computation graphs, say by allocating them on the heap instead of the stack.

**Comment by ggerganov** [comment]
> As far as I can tell, there's no way to prevent me needing such a large computation graph, because running a single graph multiple times is many times slower than simply encoding all the steps into a single graph. (Maybe because ggml messes with threads even when n_threads == 1)

The creation of a new graph with thousands of nodes and running it on every token is in the order of 100s of **microseconds** - i.e. much less than `1 ms`. Is that too slow for your use case? There are plans to optimize the graph construction and processing even further to make this overhead even more negligible in the future.

Btw, how do you create a graph for the entire sequence - don't you need to sample the probability distribution after each generated token?

In general, I didn't imagine `ggml` to support gigantic graphs, but we can reconsider if there are viable use cases. Help me understand why and how you reach millions of nodes in the graph and how you have determined that the usual approach of recreating the graph is slow.

**Comment by LoganDark** [comment]
> The creation of a new graph with thousands of nodes and running it on every token is in the order of 100s of **microseconds** - i.e. much less than `1 ms`. Is that too slow for your use case?

Yes. The oberhead of finishing one execution graph and starting another execution of the next graph is really high compared to having 2 notes in only one graph. :/

> Btw, how do you create a graph for the entire sequence - don't you need to sample the probability distribution after each generated token?

The probability distribution is derived from the state, but the state is what's used internally.

**Comment by lin72h** [comment]
> I'm on drugs right now so my ability to explain is heavily impaired, sorry lol

That's sprite! drug is needed for this kind of brain twisted task

**Comment by LoganDark** [comment]
> > I'm on drugs right now so my ability to explain is heavily impaired, sorry lol
> 
> That's sprite! drug is needed for this kind of brain twisted task

the sequel to

[![image](https://github.com/ggerganov/ggml/assets/4723091/92f8d642-1c28-42dc-95f3-84659f00dc3b)](https://github.com/saharNooby/rwkv.cpp/pull/87#issuecomment-1574851978)

**Comment by ggerganov** [comment]
OK, I see. Lets prioritize merging #264 since this would help creating customized fused ops and reduce the number of nodes and most of the overhead of the existing `ggml` primitives.

I will think about how to extend the API to support creating graphs on the heap. It shouldn't be something that we want incentivize though, since if your graph is too big, you are much better off trying to find ways to make it smaller.

**Comment by LoganDark** [comment]
> It shouldn't be something that we want incentivize though, since if your graph is too big, you are much better off trying to find ways to make it smaller.

It doesn't matter. All the nodes are stored inline in the cgraph struct. It's freaking huge (hunderds of kilobytes) even when completely empty. Move node list to heap or more cgraph to heap.

**Comment by LoganDark** [comment]
> Lets prioritize merging #264

Is there a particular reason why this is taking a while despite other PRs being merged in the meantime? Is there something I need to do first?
